src/main/frontend/fs.cljs

```diff
@@ -1,53 +1,205 @@
+          basename (last parts)
+          sub-dir (->> (butlast parts)
+                       (remove string/blank?)
+                       (string/join "/"))
+          handle-path (str "handle/"
+                           (subs dir 1)
+                           (if sub-dir
+                             (str "/" sub-dir)))
+          handle-path (if (= "/" (last handle-path))
+                        (subs handle-path 0 (dec (count handle-path)))
+                        handle-path)
+          basename-handle-path (str handle-path "/" basename)
+          file-handle-cache (get-nfs-file-handle basename-handle-path)]
+      (p/let [file-handle (or file-handle-cache (idb/get-item basename-handle-path))]
+        (when-not file-handle-cache
+          (add-nfs-file-handle! basename-handle-path file-handle))
+        (if file-handle
+          (utils/writeFile file-handle content)
+          ;; create file handle
+          (->
+           (p/let [handle (idb/get-item handle-path)]
+             (if handle
+               (p/let [file-handle (.getFileHandle ^js handle basename #js {:create true})
+                       _ (idb/set-item! basename-handle-path file-handle)]
+                 (utils/writeFile file-handle content))
+               (println "Error: directory handle not exists: " handle-path)))
+           (p/catch (fn [error]
+                      (println "Write local file failed: " {:path path})
+                      (js/console.error error)))))))
+
+    js/window.pfs
+    (js/window.pfs.writeFile (str dir "/" path) content)
+
+    :else
+    nil))
+
+(defn rename
+  [old-path new-path]
+  (cond
+    (local-db? old-path)
+    ;; create new file
+    ;; delete old file
+    (p/let [[dir basename] (util/get-dir-and-basename old-path)
+            [_ new-basename] (util/get-dir-and-basename new-path)
+            handle (idb/get-item (str "handle" old-path))
+            file (.getFile handle)
+            content (.text file)
+            _ (write-file dir new-basename content)]
+      (unlink old-path nil))
+
+    :else
+    (js/window.pfs.rename old-path new-path)))
 
 (defn stat
   [dir path]
-  (js/window.pfs.stat (str dir "/" path)))
+  (let [append-path (if path
+               (str "/" (string/replace-first path "/" ""))
+               "")]
+    (cond
+     (local-db? dir)
+     (if-let [file (get-nfs-file-handle (str "handle/"
+                                             (string/replace-first dir "/" "")
+                                             append-path))]
+       (p/let [file (.getFile file)]
+         (let [get-attr #(gobj/get file %)]
+           {:file/last-modified-at (get-attr "lastModified")
+            :file/size (get-attr "size")
+            :file/type (get-attr "type")}))
+       (p/rejected "File not exists"))
+
+     :else
+     ;; FIXME: same format
+     (js/window.pfs.stat (str dir append-path)))))
+
+(defn mkdir-if-not-exists
+  [dir]
+  (when (and dir js/window.pfs)
```

Comment:
Does this line work for local-db?

Reply:
Currently, it should works because we're waiting for the pfs is loaded, but we need to move to a `fs` protocol and only load whatever implementation the user needs.

Reply:
Yes, it should work.

Reply:
It's fixed now.